packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,752 @@
+  _16h: HTMLCanvasElement = document.createElement("canvas"); // non-stretched
+  paintWav: PaintWavFunction;
+
+  constructor(vis: Vis) {
+    super(vis);
+    this._analyser = this._vis.analyser!;
+    this._bufferLength = this._analyser.fftSize;
+    this._dataArray = new Uint8Array(this._bufferLength);
+
+    this._16h.width = 1;
+    this._16h.height = 16;
+    this._16h.setAttribute("width", "75");
+    this._16h.setAttribute("height", "16");
+
+    //* see https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio#monitoring_screen_resolution_or_zoom_level_changes
+    this._pixelRatio = window.devicePixelRatio || 1;
+
+    if (this._vis.oscStyle === "dots") {
+      this.paintWav = this.paintWavLine.bind(this);
+    } else if (this._vis.oscStyle === "solid") {
+      // does call paintWavLine but there is a check inside
+      // that changes the oscstyle accordingly
+      this.paintWav = this.paintWavLine.bind(this);
+    } else {
+      this.paintWav = this.paintWavLine.bind(this);
+    }
+  }
+
+  prepare() {
+    if (!this._ctx) {
+      console.log("ctx not set!");
+      return;
+    }
+    const vis = this._vis;
+
+    //? paint bar
+    this._bar.width = 1;
+    this._bar.height = 5;
+    this._bar.setAttribute("width", "1");
+    this._bar.setAttribute("height", "5");
+    const ctx = this._bar.getContext("2d");
+    if (ctx) {
+      for (let y = 0; y < 5; y++) {
+        ctx.fillStyle = vis.colors[18 + y];
+        ctx.fillRect(0, y, 1, y + 1);
+      }
+    }
+
+    this._ctx.imageSmoothingEnabled = false;
+    // @ts-ignore
+    this._ctx.mozImageSmoothingEnabled = false;
+    // @ts-ignore
+    this._ctx.webkitImageSmoothingEnabled = false;
+    // @ts-ignore
+    this._ctx.msImageSmoothingEnabled = false;
+  }
+
+  paintFrame() {
+    if (!this._ctx) return;
+    this._analyser.getByteTimeDomainData(this._dataArray);
+    this._dataArray = this._dataArray.slice(0, 576);
+    const bandwidth = this._dataArray.length;
+
+    const width = this._ctx!.canvas.width;
+    const height = this._ctx!.canvas.height;
+    this._ctx!.clearRect(0, 0, width, height);
+
+    const sliceWidth = Math.floor(bandwidth / width);
+
+    let y: number;
+    let colorIndex: number;
+    // Iterate over the width of the canvas in fixed 75 pixels.
+    for (let j = 0; j <= 75; j++) {
+      const amplitude = slice1st(this._dataArray, sliceWidth, j);
+      // +4 is set to off center the oscilloscope
+      // because completely centered looks a bit weird
+      if (PIXEL_DENSITY === 2){
+        [y, colorIndex] = this.rangeByAmplitude(((amplitude+4)/2)+48);
+      } else {
+        [y, colorIndex] = this.rangeByAmplitude(windowShade ? ((amplitude+4)/3)+90 : amplitude+4);
+      }
+      this.paintWav(j, y, colorIndex);
+    }
+  }
+
+  /**
+   *
+   * @param amplitude 0..255
+   * @returns xy.Y(top to bottom), colorOscIndex
+   */
+  rangeByAmplitude(amplitude: number): [number, number] {
+    // sorry about this mess
```

Comment:
Lots of magic numbers etc in this file. I don't doubt that they're needed though. Where did they come from? Are they adapted from existing code? If so, let's link back to it. Are they reverse engineered? Guessed? I think either is fine, but a comment in the relevant places would be useful so the reader knows.

Reply:
I have a commit lined up that replaced rangeByAmplitude with something more optimized, the whole section above was also simplified so that we only pass amplitude directly to paintWav and perform any post-processing there afterwards, though paintWavLine isn't really commented as that too is filled with magic numbers left and right